src/annotation/askui-annotation-step-reporter.ts

```
@@ -0,0 +1,80 @@
+import fs from "fs";
+import { Step, Reporter, Annotation, ReporterConfig } from "askui";
+import path from "path";
+import { JSDOM } from 'jsdom';
+
+export enum AnnotationLevel {
+  DISABLED = 'disabled',
+  ON_FAILURE = 'onFailure',
+  ALL = 'all',
+}
+
+export interface AnnotationReporterConfig extends ReporterConfig {
+  annotationLevel: AnnotationLevel;
+  folderPath: string;
+  fileNameTemplate: string;
+}
+
+export class AskUIAnnotationStepReporter implements Reporter {
+
+  reporterConfig: AnnotationReporterConfig = {
+    annotationLevel: AnnotationLevel.DISABLED,
+    folderPath: 'report',
+    fileNameTemplate: '_testStep_annotation',
+  }
+
+  config: ReporterConfig = {
+    withScreenshots: 'always',
+    withDetectedElements: 'always',
+  }
+
+  constructor(reporterConfig: AnnotationReporterConfig) {
+    if (reporterConfig.annotationLevel !== undefined) {
+      this.reporterConfig.annotationLevel = reporterConfig.annotationLevel;
+    }
+    if (reporterConfig.fileNameTemplate !== undefined) {
+      this.reporterConfig.fileNameTemplate = reporterConfig.fileNameTemplate;
+    }
+    if (reporterConfig.folderPath !== undefined) {
+      this.reporterConfig.folderPath = reporterConfig.folderPath;
+    }
+    if (reporterConfig.withScreenshots !== undefined) {
+      this.config.withScreenshots = reporterConfig.withScreenshots;
+    }
+    if (reporterConfig.withDetectedElements !== undefined) {
+      this.config.withDetectedElements = reporterConfig.withDetectedElements;
+    }
+  }
+
+  async onStepEnd(step: Step): Promise<void> {
+    if (( 
+          (this.reporterConfig.annotationLevel === AnnotationLevel.ON_FAILURE && step.status !== 'passed') ||
+          this.reporterConfig.annotationLevel === AnnotationLevel.ALL
+        ) &&
+        step.lastRun?.end?.screenshot !== undefined &&
+        step.lastRun?.end?.detectedElements !== undefined) {
```

They should be defined if you set the `config` correctly (you basically simply need to set `withDetectedElements` based on the `annotationLevel`). Throw an error if they are not instead of simply skipping the creation as this would indicate a bug...